# Assignment 3
### Due 9/21. Do four of five.

Sabine Segaloff

Begun: 9/16/25

In [72]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Exercise 1
- Open the NHANES (or Ames prices or college completion datasets, if you prefer)
- Find two categorical variables of interest (there are 198, and short descriptions are given in the `nhanes_meta_17_18.csv` file). Investigate their missing values (you don't have to focus on missing values for this analysis like we did with police use of force, but always be aware of how dirty the data are)
- Compute a contingency table for your categorical $X$ and $Y$
- Discuss any interesting patterns (or lack of one) that you observe

#### Open the NHANES (or Ames prices or college completion datasets, if you prefer)

In [73]:
nhanes_path = '/Users/sabinesegaloff/Code/UVA_Fall_25/Understanding_Uncertainty_5030/data/nhanes_data_17_18.csv'
ndf = pd.read_csv(nhanes_path, low_memory=False)
# Setting low_memory=False instructs Pandas to read the entire file into memory before 
# inferring data types. This allows Pandas to determine the most appropriate and consistent
# data type for each column across the entire dataset, preventing mixed-type warnings and 
# ensuring data integrity. (Doing this since without it we get a DtypeWarning saying 142
# columns have mixed types)

ndf.head()

SEQN GeneralHealthCondition  EverBreastfedOrFedBreastmilk  \
0  93703.0                    NaN                           1.0   
1  93704.0                    NaN                           1.0   
2  93705.0                   Good                           NaN   
3  93706.0              Very good                           NaN   
4  93707.0                   Good                           NaN   

   AgeStoppedBreastfeedingdays  AgeFirstFedFormuladays  \
0                        273.0                     1.0   
1                         60.0                     3.0   
2                          NaN                     NaN   
3                          NaN                     NaN   
4                          NaN                     NaN   

   AgeStoppedReceivingFormuladays  AgeStartedOtherFoodbeverage  \
0                           365.0                        152.0   
1                           365.0                        126.0   
2                             NaN                          NaN   
3                             NaN                          NaN   
4                             NaN                          NaN   

   AgeFirstFedMilkdays TypeOfMilkFirstFedWholeMilk  TypeOfMilkFirstFed2Milk  \
0                365.0       Whole or regular milk                      NaN   
1                365.0       Whole or regular milk                      NaN   
2                  NaN                         NaN                      NaN   
3                  NaN                         NaN                      NaN   
4                  NaN                         NaN                      NaN   

   ...  DaysSmokedCigsDuringPast30Days  AvgCigarettesdayDuringPast30Days  \
0  ...                             NaN                               NaN   
1  ...                             NaN                               NaN   
2  ...                             NaN                               NaN   
3  ...                             NaN                               NaN   
4  ...                             NaN                               NaN   

   TriedToQuitSmoking  TimesStoppedSmokingCigarettes  \
0                 NaN                            NaN   
1                 NaN                            NaN   
2                 NaN                            NaN   
3                 NaN                            NaN   
4                 NaN                            NaN   

  HowLongWereYouAbleToStopSmoking UnitOfMeasureDayweekmonthyear_2_SMQ  \
0                             NaN                                 NaN   
1                             NaN                                 NaN   
2                             NaN                                 NaN   
3                             NaN                                 NaN   
4                             NaN                                 NaN   

  CurrentSelfreportedHeightInches  CurrentSelfreportedWeightPounds  \
0                             NaN                              NaN   
1                             NaN                              NaN   
2                            63.0                            165.0   
3                            68.0                            145.0   
4                             NaN                              NaN   

   TriedToLoseWeightInPastYear  TimesLost10LbsOrMoreToLoseWeight  
0                          NaN                               NaN  
1                          NaN                               NaN  
2                          0.0                  11 times or more  
3                          0.0                             Never  
4                          NaN                               NaN  

[5 rows x 198 columns]

#### Find two categorical variables of interest (there are 198, and short descriptions are given in the `nhanes_meta_17_18.csv` file). Investigate their missing values (you don't have to focus on missing values for this analysis like we did with police use of force, but always be aware of how dirty the data are)

In [74]:
# Categorical variable 1: SchoolServesSchoolLunches
# Categorical variable 2: SchoolServeCompleteBreakfastEachDay

# (both are targeting people aged 4-19)

ndf[['SchoolServeCompleteBreakfastEachDay','SchoolServesSchoolLunches']]

SchoolServeCompleteBreakfastEachDay  SchoolServesSchoolLunches
0                                     NaN                        NaN
1                                     NaN                        NaN
2                                     NaN                        NaN
3                                     NaN                        NaN
4                                     1.0                        1.0
...                                   ...                        ...
8361                                  NaN                        NaN
8362                                  NaN                        NaN
8363                                  NaN                        NaN
8364                                  1.0                        1.0
8365                                  NaN                        NaN

[8366 rows x 2 columns]

In [75]:
# how many missing values for each variable?
print(ndf['SchoolServesSchoolLunches'].isna().value_counts())
print(ndf['SchoolServeCompleteBreakfastEachDay'].isna().value_counts())


SchoolServesSchoolLunches
True     6296
False    2070
Name: count, dtype: int64
SchoolServeCompleteBreakfastEachDay
True     6317
False    2049
Name: count, dtype: int64


So we are missing 6296 values for SchoolServesSchoolLunches and 6317 values for SchoolServeCompleteBreakfastEachDay

#### Compute a contingency table for your categorical $X$ and $Y$

In [76]:
pd.crosstab(ndf['SchoolServesSchoolLunches'], 
            ndf['SchoolServeCompleteBreakfastEachDay'], normalize='columns')

SchoolServeCompleteBreakfastEachDay       0.0       1.0
SchoolServesSchoolLunches                              
0.0                                  0.298077  0.003259
1.0                                  0.701923  0.996741

#### Discuss any interesting patterns (or lack of one) that you observe

Of schools that serve a complete breakfast each day, 99.6741% also serve school lunch. This is not surprising.

Also, it appears that there is at least one school serving a complete breakfast that is not serving lunch. I find that interesting. Am I reading that correctly?

## Exercise 2
- Open the NHANES dataset
- Find a categorical and numeric variable of interest (there are 198, and short descriptions are given in the `nhanes_meta_17_18.csv` file). Investigate their missing values (you don't have to focus on missing values for this analysis, but always be aware of them)
- Make descriptive tables and grouped kernel density plots to represent the variation in your numeric $Y$ conditional on your categorical $X$
- Discuss any interesting patterns (or lack of one) that you observe

#### Open the NHANES dataset

In [79]:
ndf.head()

SEQN GeneralHealthCondition  EverBreastfedOrFedBreastmilk  \
0  93703.0                    NaN                           1.0   
1  93704.0                    NaN                           1.0   
2  93705.0                   Good                           NaN   
3  93706.0              Very good                           NaN   
4  93707.0                   Good                           NaN   

   AgeStoppedBreastfeedingdays  AgeFirstFedFormuladays  \
0                        273.0                     1.0   
1                         60.0                     3.0   
2                          NaN                     NaN   
3                          NaN                     NaN   
4                          NaN                     NaN   

   AgeStoppedReceivingFormuladays  AgeStartedOtherFoodbeverage  \
0                           365.0                        152.0   
1                           365.0                        126.0   
2                             NaN                          NaN   
3                             NaN                          NaN   
4                             NaN                          NaN   

   AgeFirstFedMilkdays TypeOfMilkFirstFedWholeMilk  TypeOfMilkFirstFed2Milk  \
0                365.0       Whole or regular milk                      NaN   
1                365.0       Whole or regular milk                      NaN   
2                  NaN                         NaN                      NaN   
3                  NaN                         NaN                      NaN   
4                  NaN                         NaN                      NaN   

   ...  DaysSmokedCigsDuringPast30Days  AvgCigarettesdayDuringPast30Days  \
0  ...                             NaN                               NaN   
1  ...                             NaN                               NaN   
2  ...                             NaN                               NaN   
3  ...                             NaN                               NaN   
4  ...                             NaN                               NaN   

   TriedToQuitSmoking  TimesStoppedSmokingCigarettes  \
0                 NaN                            NaN   
1                 NaN                            NaN   
2                 NaN                            NaN   
3                 NaN                            NaN   
4                 NaN                            NaN   

  HowLongWereYouAbleToStopSmoking UnitOfMeasureDayweekmonthyear_2_SMQ  \
0                             NaN                                 NaN   
1                             NaN                                 NaN   
2                             NaN                                 NaN   
3                             NaN                                 NaN   
4                             NaN                                 NaN   

  CurrentSelfreportedHeightInches  CurrentSelfreportedWeightPounds  \
0                             NaN                              NaN   
1                             NaN                              NaN   
2                            63.0                            165.0   
3                            68.0                            145.0   
4                             NaN                              NaN   

   TriedToLoseWeightInPastYear  TimesLost10LbsOrMoreToLoseWeight  
0                          NaN                               NaN  
1                          NaN                               NaN  
2                          0.0                  11 times or more  
3                          0.0                             Never  
4                          NaN                               NaN  

[5 rows x 198 columns]

#### Find a categorical and numeric variable of interest (there are 198, and short descriptions are given in the `nhanes_meta_17_18.csv` file). Investigate their missing values (you don't have to focus on missing values for this analysis, but always be aware of them)

In [81]:
# Categorical variable: SchoolLunchFreeReducedOrFullPrice
# Numeric Variable: OfTimesweekGetSchoolLunch

# (both are targeting people aged 4-19)

ndf[['SchoolLunchFreeReducedOrFullPrice','OfTimesweekGetSchoolLunch']]

SchoolLunchFreeReducedOrFullPrice  OfTimesweekGetSchoolLunch
0                                  NaN                        NaN
1                                  NaN                        NaN
2                                  NaN                        NaN
3                                  NaN                        NaN
4                        Reduced price                        5.0
...                                ...                        ...
8361                               NaN                        NaN
8362                               NaN                        NaN
8363                               NaN                        NaN
8364                        Full price                        3.0
8365                               NaN                        NaN

[8366 rows x 2 columns]

In [83]:
# how many missing values for each variable?
print(ndf['SchoolLunchFreeReducedOrFullPrice'].isna().value_counts())
print(ndf['OfTimesweekGetSchoolLunch'].isna().value_counts())

SchoolLunchFreeReducedOrFullPrice
True     6731
False    1635
Name: count, dtype: int64
OfTimesweekGetSchoolLunch
True     6366
False    2000
Name: count, dtype: int64


So we are missing 6731 values for SchoolLunchFreeReducedOrFullPrice and 6366 values for OfTimesweekGetSchoolLunch

#### Make descriptive tables and grouped kernel density plots to represent the variation in your numeric $Y$ conditional on your categorical $X$

#### Discuss any interesting patterns (or lack of one) that you observe

## Exercise 3 
We showed that the mean and median could be discovered by minimizing various kinds of loss functions; this is what machine learning is. To make a prediction $\hat{y}(z)$ of $Y$ when $X=z$, minimize the mean squared error:
$$
MSE(\hat{y}(z)) = \dfrac{1}{N} \sum_{i=1}^N \left\lbrace y_i - \hat{y}(z) \right\rbrace^2 \frac{1}{h}k\left(\frac{z-x_i}{h}\right)
$$
Show that the solution to this problem is the LCLS/Naradaya-Watson estimator.

## Exercise 4 
- Write a class or set of functions that implement the LCLS/Naradaya-Watson estimator, using the Silverman plug-in estimate for the conditioning variable $X$ as the bandwidth.
- From one of the course data sets, find two numeric variables of interest, analyze their relationship with the the LCLS/Naradaya-Watson estimator, and discuss your results.

#### Write a class or set of functions that implement the LCLS/Naradaya-Watson estimator, using the Silverman plug-in estimate for the conditioning variable $X$ as the bandwidth.

#### From one of the course data sets, find two numeric variables of interest, analyze their relationship with the the LCLS/Naradaya-Watson estimator, and discuss your results.

## Exercise 5 
- In any of the available data sets, investigate the relationships between pairs of variables $(X,Y)$ with a scatterplot and CEF (for example, price on area)
- Is this relationship plausibly causal, or are there missing variables that might explain at least part of the relationship between your variables? These can be "conceptual" rather than "practical"; for example, 'talent' or 'grit' probably explain education outcomes, but are almost impossible to measure. We are asking whether there are hypothetical **threats to causal identification** of the effect of $X$ on $Y$.
- Explain how, regardless of the threat to causal identification, you can still use your model to predict $Y$ given $X$, as long as you don't intervene in the system to control the outcome

#### In any of the available data sets, investigate the relationships between pairs of variables $(X,Y)$ with a scatterplot and CEF (for example, price on area)

#### Is this relationship plausibly causal, or are there missing variables that might explain at least part of the relationship between your variables? These can be "conceptual" rather than "practical"; for example, 'talent' or 'grit' probably explain education outcomes, but are almost impossible to measure. We are asking whether there are hypothetical **threats to causal identification** of the effect of $X$ on $Y$.

#### Explain how, regardless of the threat to causal identification, you can still use your model to predict $Y$ given $X$, as long as you don't intervene in the system to control the outcome